File: node/src/admin/clients.ts
import type KcAdminClient from '@keycloak/keycloak-admin-client'
import type ClientRepresentation from '@keycloak/keycloak-admin-client/lib/defs/clientRepresentation.js'
import { call, requireFound } from './call.js'

/**
 * 클라이언트 CRUD 파사드. 공식 admin-client의 `clients` 리소스를 감싸며 {@link call}로 경계
 * 변환한다. `id`는 클라이언트의 내부 UUID, `clientId`는 사람이 지정한 식별자다.
 */
export class ClientsResource {
  /**
   * @internal `AdminClient`가 조립한다 — 소비자 생성 경로가 아니다. `@internal` + `stripInternal`로
   * 방출 `.d.ts`에서 지워, 하위 `KcAdminClient` 타입이 공개 표면에 오르지 않게 한다(§4).
   */
  constructor(
    private readonly kc: KcAdminClient,
    private readonly realm: string,
  ) {}

  /** 클라이언트를 생성하고 신규 클라이언트 내부 id(UUID)를 반환한다. */
  async create(representation: ClientRepresentation): Promise<string> {
    const created = await call(() =>
      this.kc.clients.create({ ...representation, realm: this.realm }),
    )
    return created.id
  }

  /** 내부 id(UUID)로 클라이언트를 조회한다. 없으면 `KeycloakNotFoundError`. */
  async get(id: string): Promise<ClientRepresentation> {
    return requireFound(
      await call(() => this.kc.clients.findOne({ id, realm: this.realm })),
      `Client not found: ${id}`,
    )
  }

  /** 사람이 지정한 `clientId`로 클라이언트를 검색한다(0..N개). */
  async findByClientId(clientId: string): Promise<ClientRepresentation[]> {
    return call(() => this.kc.clients.find({ realm: this.realm, clientId }))
  }

  async update(id: string, representation: ClientRepresentation): Promise<void> {
    await call(() => this.kc.clients.update({ id, realm: this.realm }, representation))
  }

  async delete(id: string): Promise<void> {
    await call(() => this.kc.clients.del({ id, realm: this.realm }))
  }
}

File: kotlin/src/main/kotlin/io/github/xzawed/keycloak/admin/Clients.kt
package io.github.xzawed.keycloak.admin

import jakarta.ws.rs.WebApplicationException
import org.keycloak.admin.client.CreatedResponseUtil
import org.keycloak.representations.idm.ClientRepresentation

// Clients.kt — ClientsResource: 공식 admin-client의 org.keycloak.admin.client.resource.ClientsResource를
// 감싸며 모든 호출을 adminCall(AdminClient.kt)로 경계 변환한다(Java ClientsResource 동형).
//
// get()은 UsersResource.get()과 동일한 정책을 따른다: 대상이 없으면 KeycloakAdminException.NotFound를
// 그대로 전파한다.
public class ClientsResource internal constructor(
    private val delegate: org.keycloak.admin.client.resource.ClientsResource,
) {
    /** 클라이언트 생성 후 응답 Location 헤더에서 신규 클라이언트 id를 추출해 반환한다. */
    public suspend fun create(representation: ClientRepresentation): String =
        adminCall {
            val response = delegate.create(representation)
            try {
                CreatedResponseUtil.getCreatedId(response)
            } finally {
                response.close()
            }
        }

    public suspend fun get(id: String): ClientRepresentation = adminCall { delegate.get(id).toRepresentation() }

    public suspend fun findByClientId(clientId: String): List<ClientRepresentation> = adminCall { delegate.findByClientId(clientId) }

    public suspend fun update(
        id: String,
        representation: ClientRepresentation,
    ) {
        adminCall { delegate.get(id).update(representation) }
    }

    /**
     * admin-client의 `delegate.delete(id)`는 [jakarta.ws.rs.core.Response]를 반환하는 JAX-RS 프록시
     * 메서드라 4xx/5xx 응답에서도 예외를 던지지 않는다(void 반환 메서드만 자동 throw한다). 상태 코드를
     * 직접 검사해 실패 시 [WebApplicationException]을 던져 [adminCall] 경계에서 SDK 예외로 변환한다.
     */
    public suspend fun delete(id: String) {
        adminCall {
            val response = delegate.delete(id)
            try {
                if (response.status >= 400) {
                    throw WebApplicationException(response)
                }
            } finally {
                response.close()
            }
        }
    }
}
